File: python/mirage/core/lancedb/scope.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from dataclasses import dataclass, field
from enum import Enum

from mirage.resource.lancedb.config import LanceDBConfig
from mirage.types import PathSpec


class ScopeLevel(str, Enum):
    ROOT = "root"
    GROUP_DIR = "group_dir"
    SEARCH_DIR = "search_dir"
    SEARCH_RESULTS = "search_results"
    ROW = "row"
    UNKNOWN = "unknown"


@dataclass
class LanceDBScope:
    level: ScopeLevel
    table: str | None = None
    filters: dict[str, str] = field(default_factory=dict)
    query: str | None = None
    row_id: str | None = None
    blob: bool = False
    resource_path: str = "/"


def _parse_row_file(name: str, config: LanceDBConfig) -> tuple[str, bool] | None:
    if name.endswith(".md"):
        return name[:-len(".md")], False
    if config.blob_column:
        suffix = "." + config.blob_ext
        if name.endswith(suffix):
            return name[:-len(suffix)], True
    return None


def detect_scope(path, config: LanceDBConfig) -> LanceDBScope:
    raw = path.strip_prefix if isinstance(path, PathSpec) else path
    key = raw.strip("/")
    segs = key.split("/") if key else []

    if config.table:
        table = config.table
        rest = segs
    else:
        if not segs:
            return LanceDBScope(level=ScopeLevel.ROOT, resource_path=raw)
        table = segs[0]
        rest = segs[1:]

    if rest and rest[0] == config.search_dir:
        qparts = rest[1:]
        if not qparts:
            return LanceDBScope(level=ScopeLevel.SEARCH_DIR,
                                table=table,
                                resource_path=raw)
        query = qparts[0]
        if len(qparts) == 1:
            return LanceDBScope(level=ScopeLevel.SEARCH_RESULTS,
                                table=table,
                                query=query,
                                resource_path=raw)
        if len(qparts) == 2:
            parsed = _parse_row_file(qparts[1], config)
            if parsed is not None:
                return LanceDBScope(level=ScopeLevel.ROW,
                                    table=table,
                                    query=query,
                                    row_id=parsed[0],
                                    blob=parsed[1],
                                    resource_path=raw)
        return LanceDBScope(level=ScopeLevel.UNKNOWN, resource_path=raw)

    gb = config.group_by
    n = len(gb)

    if len(rest) <= n:
        filters = {gb[i]: rest[i] for i in range(len(rest))}
        return LanceDBScope(level=ScopeLevel.GROUP_DIR,
                            table=table,
                            filters=filters,
                            resource_path=raw)

    if len(rest) == n + 1:
        filters = {gb[i]: rest[i] for i in range(n)}
        parsed = _parse_row_file(rest[n], config)
        if parsed is not None:
            return LanceDBScope(level=ScopeLevel.ROW,
                                table=table,
                                filters=filters,
                                row_id=parsed[0],
                                blob=parsed[1],
                                resource_path=raw)

    return LanceDBScope(level=ScopeLevel.UNKNOWN, resource_path=raw)

File: typescript/packages/core/src/core/lancedb/scope.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { LanceDBConfigResolved } from '../../resource/lancedb/config.ts'
import { PathSpec } from '../../types.ts'
import { stripSlash } from '../../utils/slash.ts'

export const ScopeLevel = Object.freeze({
  ROOT: 'root',
  GROUP_DIR: 'group_dir',
  ROW: 'row',
  UNKNOWN: 'unknown',
} as const)

export type ScopeLevel = (typeof ScopeLevel)[keyof typeof ScopeLevel]

export interface LanceDBScope {
  level: ScopeLevel
  table: string | null
  filters: Record<string, string>
  rowId: string | null
  blob: boolean
  resourcePath: string
}

function parseRowFile(name: string, config: LanceDBConfigResolved): [string, boolean] | null {
  if (name.endsWith('.md')) return [name.slice(0, -'.md'.length), false]
  if (config.blobColumn !== null) {
    const suffix = `.${config.blobExt}`
    if (name.endsWith(suffix)) return [name.slice(0, -suffix.length), true]
  }
  return null
}

function make(
  level: ScopeLevel,
  resourcePath: string,
  over: Partial<LanceDBScope> = {},
): LanceDBScope {
  return {
    level,
    table: over.table ?? null,
    filters: over.filters ?? {},
    rowId: over.rowId ?? null,
    blob: over.blob ?? false,
    resourcePath,
  }
}

export function detectScope(path: PathSpec | string, config: LanceDBConfigResolved): LanceDBScope {
  const raw = path instanceof PathSpec ? path.stripPrefix : path
  const key = stripSlash(raw)
  const segs = key === '' ? [] : key.split('/')

  let table: string
  let rest: string[]
  if (config.table !== null) {
    table = config.table
    rest = segs
  } else {
    if (segs.length === 0) return make(ScopeLevel.ROOT, raw)
    table = segs[0] ?? ''
    rest = segs.slice(1)
  }

  const gb = config.groupBy
  const n = gb.length

  if (rest.length <= n) {
    const filters: Record<string, string> = {}
    for (let i = 0; i < rest.length; i++) filters[gb[i] ?? ''] = rest[i] ?? ''
    return make(ScopeLevel.GROUP_DIR, raw, { table, filters })
  }

  if (rest.length === n + 1) {
    const filters: Record<string, string> = {}
    for (let i = 0; i < n; i++) filters[gb[i] ?? ''] = rest[i] ?? ''
    const parsed = parseRowFile(rest[n] ?? '', config)
    if (parsed !== null) {
      return make(ScopeLevel.ROW, raw, { table, filters, rowId: parsed[0], blob: parsed[1] })
    }
  }

  return make(ScopeLevel.UNKNOWN, raw)
}
